[chore] add derive macro tests

r3stl355 opened the pull request:
## What changes are proposed in this pull request?
<!--
Please clarify what changes you are proposing and why the changes are needed.
The purpose of this section is to outline the changes, why they are needed, and how this PR fixes the issue.
If the reason for the change is already explained clearly in an issue, then it does not need to be restated here.
  1. If you propose a new API or feature, clarify the use case for a new API or feature.
  2. If you fix a bug, you can clarify why it is a bug.
-->
A simplified version of [PR429](https://github.com/delta-incubator/delta-kernel-rs/pull/429) (to be closed/abandoned) using a [doctest trick](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#include-items-only-when-collecting-doctests).

- I had to make a small change to the macro code itself, changing `crate::` to`delta_kernel::` in the generated code.
- Used existing `developer-visibility` feature instead of adding a new one as it seemed to fit the purpose
- Added an extra command to `justfile` to run the new tests as part of existing testing process

<!--
Uncomment this section if there are any changes affecting public APIs:
### This PR affects the following public APIs

If there are breaking changes, please ensure the `breaking-changes` label gets added by CI, and describe why the changes are needed.

Note that _new_ public APIs are not considered breaking.
-->


## How was this change tested?
All tests pass, including the newly added

scovich commented on kernel/src/lib.rs:
```diff
@@ -415,3 +418,70 @@ pub trait Engine: AsAny {
 }
+
+#[cfg(doctest)]
+mod doc_tests {
```
Another question: For cases where compilation and execution succeed, should we use normal tests instead of doc tests? We only _need_ doc tests for cases that expect compilation failure, right?

r3stl355 replied:
yes, exactly, that's why I only wanted to test for failure. I kinda shot myself in the foot by adding successful compiles :)

r3stl355 replied:
How about this: I am going to leave only one successful macro compile doctest just to assure the methodology is working. Successful tests can be implemented as normal integration tests. What say you @scovich?

scovich replied:
Sounds great, thanks. If you decide not to tackle those integration tests, can you file a tracking ticket for the TODO?

r3stl355 commented on kernel/src/lib.rs:
````diff
@@ -415,3 +418,70 @@ pub trait Engine: AsAny {
+
+    /// ```
+    /// # use delta_kernel_derive::Schema;
+    /// # use std::collections::HashMap;
+    /// #[derive(Schema)]
+    /// pub struct WithAttributedField {
+    ///     #[drop_null_container_values]
+    ///     map_field: HashMap<String, String>,
````
interesting, I'll check this later

r3stl355 replied:
I think I found a fix, will add to this PR but will not add more tests, those should be done as part of #526

r3stl355 commented on kernel/src/lib.rs:
```diff
@@ -415,3 +418,70 @@ pub trait Engine: AsAny {
+    /// # use syn::Token;
+    /// #[derive(Schema)]
+    /// pub struct WithInvalidFieldType {
+    ///     token: Token![struct],
```
It is the latter, I am trying to test all the code paths in the macro. Actually, there is one path I cannot trigger. For that
I need to craft something with `PathArguments::Parenthesized` but I run out of ideas.

I asked around [on StackOverlow here](https://stackoverflow.com/questions/79123128/how-to-declare-synpatharguments-parenthesized-field-in-structure-in-rust]) but didn't get anything solid.

 This one:
```
_ => Error::new(segment.arguments.span(), "Can only handle <> type path args").to_compile_error()
```

scovich replied:
> > What is `Token![struct]`? Is it intentionally introducing a rust syntax error to see if the macro handles it gracefully? Or does it somehow produce valid rust syntax that we expect the macro cannot handle?
>
> It is the latter, I am trying to test all the code paths in the macro. 

Maybe silly question, but: If it's valid rust syntax, why can't we just type it in directly? Why do we need to inject it with syn?

r3stl355 replied:
Typing it directly will not trigger the compile error in the macro I was trying to trigger (it will create another compile error but not the one I am trying to cause), this line in the `derive-macros/src/lib.rs`
```
_ => Error::new(field.span(), format!("Can't handle type: {:?}", field.ty)).to_compile_error()
```

Token! was the first type I found that I found which would help me trigger that and therefore I used it. Now I just realised that `syn` is actually not a `dev-dependency` of the project so it has to be added it to `Cargo.toml`. However, I'll try to see if I can find a type which I can use without adding `syn`

r3stl355 replied:
This is actually a reason why the previous version, one using `trybuild`, was a bit better IMO as it would capture the exact error whereas here I'm only causing the compile error without know exactly which one is raised

scovich replied:
Yes... but the downside is every rustc update has the potential to change the error text, ie https://github.com/delta-io/delta-kernel-rs/pull/346 (the FFI `Handle` API currently relies on trybuild tests, because I hadn't learned the export+visibility tricks that would allow doc tests to work). 

On the other hand, codecov doesn't seem to "count" executable doctests as coverage for some reason, while presumably it would count trybuild tests, to the extent they exercise any kernel code. Maybe that's just a problem with our CI, I don't know?

r3stl355 replied:
Totally agree, neither solution is ideal imo but the one with doctests requires less effort to create and maintain so I'll stick with that (this PR) but will stick to `compile_fail` scenarios only (maybe have one test that succeeds just to assure the methodology is still working)

r3stl355 commented on kernel/src/lib.rs:
```diff
@@ -415,3 +418,70 @@ pub trait Engine: AsAny {
+    /// #[derive(Schema)]
+    /// pub struct WithFields {
+    ///     some_name: String,
+    /// }
```
if you are planning to merge your PR before the end of weekend then I better wait, I don't think I'll have time before the weekend to work on this

r3stl355 replied:
I'll leave this out of this PR and implement separately if the answer to the below comment is Yes (about moving successful compiles to normal tests)

zachschuermann replied:
can we create a follow-up issue to track this?

r3stl355 replied:
already done, tracked in #526

r3stl355 commented on kernel/src/lib.rs:
```diff
@@ -415,3 +418,70 @@ pub trait Engine: AsAny {
 }
+
+#[cfg(doctest)]
+mod doc_tests {
```
I'll leave this out of this PR and implement separately if the answer to the below comment is Yes (about moving successful compiles to normal tests)

scovich commented on kernel/src/lib.rs:
```diff
@@ -50,6 +50,8 @@
     rust_2021_compatibility
 )]
 
+/// This allows the derive macro to refer to types in this crate using full crate name, instead of using `crate::`,
+/// which seem to be the only way to to test the macro generated code and macro compile failures
```
This `extern crate` declaration allows the macro to reliably refer to `delta_kernel::schema::DataType` no matter which crate invokes it. Without that, `delta_kernel` cannot invoke the macro because `delta_kernel` is an unknown crate identifier (you have to use `crate` instead). We could make the macro use `crate::schema::DataType` instead, but then the macro is useless outside the `delta_kernel` crate.

scovich commented on derive-macros/src/lib.rs:
```diff
@@ -124,9 +124,9 @@ fn gen_schema_fields(data: &Data) -> TokenStream {
                             ).to_compile_error()
                         }
                     }
-                    quote_spanned! { field.span() => #(#type_path_quoted),* get_nullable_container_struct_field(stringify!(#name))}
```
Wow, subtle! So a path like `std::collections::HashMap<String, String>` was being emitted as `std::,collections::,HashMap<String, String>`?

r3stl355 replied:
yep

zachschuermann replied:
how did this compile before then haha?

r3stl355 replied:
I think it was never used with fully qualified types (or partially qualified types that contained more that one part separated by ::). `#(),*` acts kinda like `','.join` in Python - no comma is inserted if there is only one element in the list

r3stl355 replied:
Actually there is another fix few lines above I forgot to add, will do that soon

scovich reviewed (approved):
LGTM

zachschuermann commented on kernel/src/actions/mod.rs:
```diff
@@ -24,7 +24,11 @@ use serde::{Deserialize, Serialize};
 pub mod deletion_vector;
 pub mod set_transaction;
 
+#[cfg(feature = "developer-visibility")]
+pub mod schemas;
```
I suppose we need to make everything `pub` for the doctest stuff to work? I understand the need but want to flag as a downside..

scovich replied:
We _could_ check for `doctest` instead of `developer-features` -- but as @r3stl355 noted already, we use the latter already in so many places that this is the less invasive approach?
(alternatively, we could pursue the `doctest` feature gating, and use cfg any for the cases that overlap with `developer-features`)?

zachschuermann commented on kernel/src/lib.rs:
```diff
@@ -50,7 +50,12 @@
     rust_2021_compatibility
 )]
 
+/// This allows the derive macro to refer to types in this crate using full crate name, instead of using `crate::`,
+/// which seem to be the only way to to test the macro generated code and macro compile failures
```
```suggestion
/// This `extern crate` declaration allows the macro to reliably refer to
/// `delta_kernel::schema::DataType` no matter which crate invokes it. Without that, `delta_kernel`
/// cannot invoke the macro because `delta_kernel` is an unknown crate identifier (you have to use
/// `crate` instead). We could make the macro use `crate::schema::DataType` instead, but then the
/// macro is useless outside the `delta_kernel` crate.
```

zachschuermann reviewed (commented):
left a few questions!

r3stl355 commented:
I added a second part that fixes an error in macro when fully-qulified multi-part types are used (forgot to push this earlier). Also added a passing docs test to verify, this test needs to be moved to normal test as part of #526

scovich reviewed (approved):
LGTM, thanks!

zachschuermann commented on kernel/src/lib.rs:
```diff
@@ -416,3 +424,59 @@ pub trait Engine: AsAny {
     fn get_parquet_handler(&self) -> Arc<dyn ParquetHandler>;
 }
+
+#[cfg(doctest)]
+mod doc_tests {
```
can we add a comment on the requirement to use doctests here?

r3stl355 replied:
I added a comment, lmk if you had something else in mind

zachschuermann reviewed (approved):
hey @r3stl355 are you able to pick this back up? just added a quick comment and would love to have this land :)

r3stl355 commented:
Made a bit of a mess and had to force-push, hope that's alright. Should be good now

scovich commented on kernel/src/actions/mod.rs:
```diff
@@ -29,6 +29,9 @@ use serde::{Deserialize, Serialize};
 pub mod deletion_vector;
 pub mod set_transaction;
 
+#[cfg(feature = "internal-api")]
+pub mod schemas;
+#[cfg(not(feature = "internal-api"))]
 pub(crate) mod schemas;
```
See L36 below, we recently added a new way to handle internal-api mod:
```suggestion
internal_mod!(pub(crate) mod schemas);
```

r3stl355 replied:
Ahh, yes, was aware of the new way to handle inline mods but missed this occurrence, fixed it

scovich reviewed (approved):
Very nice! Just one simplification left and should be ready for merge

scovich commented:
Hmm @zachschuermann it looks like something went wrong with the doctest+internal API combo? I thought we had set the CI to enable internal API while compiling doctests?

r3stl355 commented:
Also, strangely, if I run just `cargo test --doc` on my machine without specifying any features the tests pass.

scovich commented:
> where does this one sit with all the recent changes, is it still relevant?

AFAIK, this is still important. In particular, it adds the ability to create working doctests for derive macros, which other recent PR would benefit from. I'll let @zachschuermann weigh in on whether/how this PR may need to expand tho.

zachschuermann commented:
Hi @r3stl355! Yes this is even _more_ relevant after #830 landed :) - see the PR description over there for some overlapping derive macro test work that we need.

I opened another issue that basically just links to this PR and my PR: https://github.com/delta-io/delta-kernel-rs/issues/991

I believe the remaining work would just be to (1) update this PR to work with the original stuff it intended to test and (2) add to this or open a new PR to expand the testing to my new `#[derive(IntoEngineData)]` macro!

r3stl355 commented:
Cool, thanks @scovich  & @zachschuermann, I'll see what I can come up with in the next few days

r3stl355 commented:
Hi @zachschuermann, I updated the PR to work with latest changes. I suggest creating a different one for `#[derive(IntoEngineData)]` macro, otherwise I'll never get this one over the fence :)

zachschuermann commented on kernel/src/schema/derive_macro_utils.rs:
```diff
@@ -7,6 +7,11 @@ use crate::schema::{ArrayType, DataType, MapType, StructField, ToSchema};
 
 /// Converts a type to a [`DataType`]. Implemented for the primitive types and automatically derived
 /// for all types that implement [`ToSchema`].
+#[cfg(feature = "internal-api")]
+pub trait ToDataType {
+    fn to_data_type() -> DataType;
+}
+#[cfg(not(feature = "internal-api"))]
```
does directly doing `#[internal_api]` not work?

zachschuermann replied:
sorry should have been more specific - you can see many other instances of us marking things as `#[internal_api]` which is a decorator that will be less verbose than the manual `cfg(...)` - I _think_ it should work here, but not on the module declaration

r3stl355 replied:
Ahh, got it, will review

r3stl355 replied:
Alright, it worked nicely, changed

r3stl355 commented:
Thank you for merging @zachschuermann, I'll start looking into the other macro tests
